physx/src/query.rs

```diff
@@ -0,0 +1,290 @@
+        })
+    }
+
+    unsafe fn get_raycast_hit(&mut self) -> Option<physx_sys::PxRaycastHit> {
+        Some(self.0.obj.block)
+    }
+
+    unsafe extern "C" fn process_touches(
+        _buffer: *const physx_sys::PxRaycastHit,
+        _num_hits: u32,
+        _user_data: *const c_void,
+    ) -> bool {
+        // Unused by this impl
+        false
+    }
+
+    unsafe extern "C" fn finalize_query(_user_data: *const c_void) {
+        // Unused by this impl
+    }
+
+    unsafe fn into_px(&mut self) -> *mut physx_sys::PxRaycastCallback {
+        self.0.obj.as_mut_ptr()
+    }
+}
+
+/// A trait for creating custom raycast callbacks for PhysX.
+pub unsafe trait RaycastCallback: Sized {
+    unsafe fn new() -> Self;
+    unsafe fn get_raycast_hit(&mut self) -> Option<physx_sys::PxRaycastHit>;
+
+    /// # Safety
+    /// Must not panic.
+    unsafe extern "C" fn process_touches(
+        buffer: *const physx_sys::PxRaycastHit,
+        num_hits: u32,
+        user_data: *const c_void,
+    ) -> bool;
+
+    /// # Safety
+    /// Must not panic.
+    unsafe extern "C" fn finalize_query(user_data: *const c_void);
+
+    /// # Safety
+    /// Do not override this method.
+    unsafe fn into_px(&mut self) -> *mut physx_sys::PxRaycastCallback {
+        create_raycast_callback(
+            Self::process_touches,
+            Self::finalize_query,
+            Box::into_raw(Box::new(self)) as *mut c_void,
+        )
+    }
+}
+
+#[repr(transparent)]
+pub struct PxSweepCallback {
+    pub obj: Owner<physx_sys::PxSweepCallback>,
+}
+
+#[repr(transparent)]
+pub struct PxSweepBuffer(PxSweepCallback);
+
+unsafe impl Class<PxSweepCallback> for PxSweepBuffer {
+    fn as_ptr(&self) -> *const PxSweepCallback {
+        self.0.obj.as_ptr() as *const _ as *const _
+    }
+    fn as_mut_ptr(&mut self) -> *mut PxSweepCallback {
+        self.0.obj.as_mut_ptr() as *mut _ as *mut _
+    }
+}
+
+unsafe impl Class<physx_sys::PxSweepCallback> for PxSweepBuffer {
+    fn as_ptr(&self) -> *const physx_sys::PxSweepCallback {
+        self.0.obj.as_ptr() as *const _ as *const _
+    }
+    fn as_mut_ptr(&mut self) -> *mut physx_sys::PxSweepCallback {
+        self.0.obj.as_mut_ptr() as *mut _ as *mut _
+    }
+}
+
+unsafe impl SweepCallback for PxSweepBuffer {
+    unsafe fn new() -> Self {
+        Self(PxSweepCallback {
+            obj: Owner::from_raw(create_sweep_buffer()).unwrap(),
+        })
+    }
+
+    unsafe fn get_sweep_hit(&mut self) -> Option<physx_sys::PxSweepHit> {
+        Some(self.0.obj.block)
+    }
+
+    unsafe extern "C" fn process_touches(
+        _buffer: *const physx_sys::PxSweepHit,
+        _num_hits: u32,
+        _user_data: *const c_void,
+    ) -> bool {
+        // Unused by this impl
+        false
+    }
+
+    unsafe extern "C" fn finalize_query(_user_data: *const c_void) {
+        // Unused by this impl
+    }
+
+    unsafe fn into_px(&mut self) -> *mut physx_sys::PxSweepCallback {
```

Comment:
semantically, I think this should be `as_raw` or `as_mut_ptr`? `into` is often used for sinking calls, which this isn't.